### DISCUSSION TITLE
Give entities custom IDs (collections, documents, functions, etc)

### DISCUSSION THREAD

**Discussion by alexweininger** [appwrite/appwrite#1300] [Ideas]
This was discussed in the Discord server, and I'm bringing the discussion here for better tracking.

## Proposal

It would be helpful if we could create collections and documents with custom IDs. Many document based databases (MongoDB for example) allow for this and it makes for really great DX.

## Challenges and discussion points

Some initial challenges and things to discuss brought up by @eldadfux were:

1. How the API is going to look?
2. Which entities (collections, documents, files, users, functions, etc...) will be entitled to custom ID and how can we create a consistent API for all of them

**Comment by Meldiron** [comment]
I think it would be really useful, in my scenario for functions. We separate development and production appwrite servers and there is no way to have the same ID on both. This results in additional steps both in the deployment phase and usage on the frontend.

What prevents functionName from being used to reference a function instead of ID? Would custom ID counter this problem?

**Comment by alexweininger** [comment]
> Which entities (collections, documents, files, users, functions, etc...) will be entitled to custom ID and how can we create a consistent API for all of them


The most important entities to me are collections and documents. I also think that files and functions deserve custom IDs as well.

### Collections

Making the case as to why collections should have custom IDs is the easiest in my opinion. My reasoning is that when you're referencing the ID of a collection, you almost always know the ID beforehand. It's always a constant that I have to copy to my code. E.g. `const colorsCollectionId = 'yyy'`. 

This would also be familiar to how MongoDB works as with MongoDB you create collections and give them names. And then you reference these collections by name.

I also think the same reasoning can be applied to functions.

### Documents

There is less of an argument for allowing documents to have a custom ID. The only reason that I have is that I could use a userId as the documentId. However, this feature could be provided by the upcoming database changes that will allow specifying custom unique indexes.

**Comment by eldadfux** [comment]
As mentioned on Discord, I think the team is pretty much aligned that this should be supported `Custom IDs` at some level. I think the API look and feel both for REST, SDKs and soon to be GraphQL should have a consisted experience (current and future). Understanding which resources are entitled to this feature is also tricky. I can definitely imagine use cases where this is crucial for both files and documents and can improve dev experience bu a lot.

That said, a crappy API for managing them will destroy and improvement the dev experience will get from this feature  😅

**Comment by eldadfux** [comment]
An official RFC is available here: https://github.com/appwrite/rfc/pull/26 - please post your feedback
